lib/volcanic/authenticator/v1/scope.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+module Volcanic::Authenticator
+  module V1
+    # Given a scope will contain optional wildcards or exact values
+    # we need to break out the scope into core parts to build a regex
+    # then apply it to the vrn.
+    class Scope
+      def initialize(scope)
+        @scope = scope
+      end
+
+      def valid?(vrn)
+        !vrn.match(verify_regex).nil?
+      end
+
+      attr_accessor :scope
+
+      private
+
+      # vrn:stack:dataset:model:id?qualifier
+      # Returns
+      # => Array["stack", "dataset", "model", "id", "qualifier"]
+      #
+      # Example
+      # scope = "vrn:local:-1:identity/*"
+      # scope.match(regex).captures
+      # => ["local", "-1", "identity", "*", nil]
+      #
+      # scope = "vrn:local:-1:identity/*?user_id=1"
+      # scope.match(regex).captures
+      # => ["local", "-1", "identity", "*", "user_id=1"]
+      def scope_parts
+        @scope_parts ||= scope.match(%r{^vrn:(.+):(.+):(.+)\/([\d\*])(\?.+)?}).captures
+      end
+
+      def verify_regex
+        %r{^vrn:(.+):(.+):(.+)\/(#{int_or_wild}{1}$)(\?.+)?}
```

Comment:
this is taking me way too long to figure out what its doing... and I don't think it is something that can really be done safely with a regex...

There are a couple of things in here which are important (and unhandled) such as different datasets or different stack IDs, or wildcards in those.

For simplicity, I would suggest using the regex matching only for string validation (is it a valid VRN?) and breaking up the component parts. 

Consider below (thrown together with very limited testing):

```ruby
class Scope
  MATCH_RE = %r{^vrn:([a-z\d\-\*]+):([a-z\d\-\*]+):([a-z][a-z-]*)(?:\/([\da-z\.-]+|\*))?(?:\?(\S+))?$}

  class << self
    def parse(value)
      captures = value.match(Scope::MATCH_RE).captures
      new(stack_id: captures[0],
          dataset_id: captures[1],
          resource: captures[2],
          resource_id: captures[3],
          qualifiers: captures[4])
    end
  end

  attr_reader :stack_id, :dataset_id, :resource, :resource_id, :qualifiers

  def initialize(stack_id: '*', dataset_id: '*', resource: '*', resource_id: nil, qualifiers: nil)
    @stack_id = stack_id
    @dataset_id = dataset_id
    @resource = resource
    @resource_id = resource_id
    @qualifiers = qualifiers
  end

  def include?(other)
    other = Scope.parse(other) if other.is_a?(String)
    return false unless other.is_a? Scope

    stack_included?(other) &&
      dataset_included?(other) &&
      resource_included?(other) &&
      resource_id_included?(other)
  end

   def to_s
     base = "vrn:#{stack_id}:#{dataset_id}:#{resource}"
     base += "/#{resource_id}" if resource_id
     base += "?#{qualifiers}" if qualifiers
     base
   end

   private

   def stack_included?(other)
     stack_id == '*' || stack_id == other.stack_id
   end

   def dataset_included?(other)
     dataset_id == '*' || dataset_id == other.dataset_id
   end

   def resource_included?(other)
     resource == '*' || resource == other.resource
   end

   def resource_id_included?(other)
     resource_id == '*' || resource_id == other.resource_id
   end
end
```